gateway/src/node/lp_listener/handshake.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2025 - Nym Technologies SA <contact@nymtech.net>
+// SPDX-License-Identifier: GPL-3.0-only
+
+use crate::error::GatewayError;
+use nym_lp::{
+    state_machine::{LpAction, LpInput, LpStateMachine},
+    LpPacket, LpSession,
+};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::TcpStream;
+use tracing::*;
+
+/// Wrapper around the nym-lp state machine for gateway-side LP connections
+pub struct LpGatewayHandshake {
+    state_machine: LpStateMachine,
+}
+
+impl LpGatewayHandshake {
+    /// Create a new responder (gateway side) handshake
+    ///
+    /// # Arguments
+    /// * `gateway_ed25519_keypair` - Gateway's Ed25519 identity keypair (for PSQ auth and X25519 derivation)
+    /// * `client_ed25519_public_key` - Client's Ed25519 public key (from ClientHello)
+    /// * `salt` - Salt from ClientHello (for PSK derivation)
+    pub fn new_responder(
+        gateway_ed25519_keypair: (
+            &nym_crypto::asymmetric::ed25519::PrivateKey,
+            &nym_crypto::asymmetric::ed25519::PublicKey,
+        ),
+        client_ed25519_public_key: &nym_crypto::asymmetric::ed25519::PublicKey,
+        salt: &[u8; 32],
+    ) -> Result<Self, GatewayError> {
+        let state_machine = LpStateMachine::new(
+            false, // responder
+            gateway_ed25519_keypair,
+            client_ed25519_public_key,
+            salt,
+        )
+        .map_err(|e| {
+            GatewayError::LpHandshakeError(format!("Failed to create state machine: {}", e))
+        })?;
+
+        Ok(Self { state_machine })
+    }
+
+    /// Complete the handshake and return the established session
+    pub async fn complete(mut self, stream: &mut TcpStream) -> Result<LpSession, GatewayError> {
+        debug!("Starting LP handshake as responder");
+
+        // Start the handshake
+        if let Some(action) = self.state_machine.process_input(LpInput::StartHandshake) {
+            match action {
+                Ok(LpAction::SendPacket(packet)) => {
+                    self.send_packet(stream, &packet).await?;
+                }
+                Ok(_) => {
+                    // Unexpected action at this stage
+                    return Err(GatewayError::LpHandshakeError(
+                        "Unexpected action at handshake start".to_string(),
+                    ));
+                }
+                Err(e) => {
+                    return Err(GatewayError::LpHandshakeError(format!(
+                        "Failed to start handshake: {}",
+                        e
+                    )));
+                }
+            }
+        }
+
+        // Continue handshake until complete
+        loop {
+            // Read incoming packet
+            let packet = self.receive_packet(stream).await?;
```

Comment:
this is one instance of many, but we're not checking for shutdown - what if we end up waiting until the timeout and block the whole process?